packages/core/lib/queues/HandlerSpy.ts

```diff
@@ -0,0 +1,115 @@
+import { Fifo } from 'toad-cache'
+
+import type { MessageProcessingResult } from '../types/MessageQueueTypes'
+
+export type HandlerSpyParams = {
+  bufferSize?: number
+  messageIdField?: string
+}
+
+export type SpyResult<MessagePayloadSchemas extends object> = {
+  message: MessagePayloadSchemas
+  processingResult: MessageProcessingResult
+}
+
+type SpyResultCacheEntry<MessagePayloadSchemas extends object> = {
+  value: SpyResult<MessagePayloadSchemas>
+}
+
+type SpyPromiseMetadata<MessagePayloadSchemas extends object> = {
+  fields: Partial<MessagePayloadSchemas>
+  processingResult?: MessageProcessingResult
+  promise: Promise<SpyResult<MessagePayloadSchemas>>
+  resolve: (
+    value: SpyResult<MessagePayloadSchemas> | PromiseLike<SpyResult<MessagePayloadSchemas>>,
+  ) => void
+}
+
+export class HandlerSpy<MessagePayloadSchemas extends object> {
+  private readonly messageBuffer: Fifo<SpyResult<MessagePayloadSchemas>>
+  private readonly messageIdField: string
+  private readonly spyPromises: SpyPromiseMetadata<MessagePayloadSchemas>[]
+
+  constructor(params: HandlerSpyParams = {}) {
+    this.messageBuffer = new Fifo(params.bufferSize ?? 100)
+    this.messageIdField = params.messageIdField ?? 'id'
+    this.spyPromises = []
+  }
+
+  private messageMatchesFilter(
+    spyResult: SpyResult<MessagePayloadSchemas>,
+    fields: Partial<MessagePayloadSchemas>,
+    processingResult?: MessageProcessingResult,
+  ) {
+    return (
+      Object.entries(fields).every(([key, value]) => {
+        // @ts-ignore
+        return spyResult.message[key] === value
+      }) &&
+      (!processingResult || spyResult.processingResult === processingResult)
+    )
+  }
+
+  waitForEvent(
+    fields: Partial<MessagePayloadSchemas>,
+    processingResult?: MessageProcessingResult,
+  ): Promise<SpyResult<MessagePayloadSchemas>> {
+    // @ts-ignore
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
```

Comment:
this needs to be fixed upstream in caching library, it currently types `items` as `any` :-/